test_conformance/extensions/cl_khr_semaphore/test_semaphores_cross_queue.cpp

```diff
@@ -0,0 +1,368 @@
+
+        if (single_queue)
+        {
+            consumer_queue = producer_queue;
+        }
+        else
+        {
+            consumer_queue = clCreateCommandQueue(
+                context, device, CL_QUEUE_OUT_OF_ORDER_EXEC_MODE_ENABLE,
+                &error);
+            test_error(error, "Could not create command queue");
+        }
+
+        // Create semaphore
+        cl_semaphore_properties_khr sema_props[] = {
+            static_cast<cl_semaphore_properties_khr>(CL_SEMAPHORE_TYPE_KHR),
+            static_cast<cl_semaphore_properties_khr>(
+                CL_SEMAPHORE_TYPE_BINARY_KHR),
+            0
+        };
+        semaphore =
+            clCreateSemaphoreWithPropertiesKHR(context, sema_props, &error);
+        test_error(error, "Could not create semaphore");
+
+        // create memory resources
+        in_mem_A = clCreateBuffer(context, CL_MEM_READ_ONLY,
+                                  sizeof(cl_int) * num_elems, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        in_mem_B = clCreateBuffer(context, CL_MEM_READ_ONLY,
+                                  sizeof(cl_int) * num_elems, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+
+        out_mem_A = clCreateBuffer(context, CL_MEM_WRITE_ONLY,
+                                   sizeof(cl_int) * num_elems, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        out_mem_B = clCreateBuffer(context, CL_MEM_WRITE_ONLY,
+                                   sizeof(cl_int) * num_elems, nullptr, &error);
+        test_error(error, "clCreateBuffer failed");
+
+        error = clSetKernelArg(kernel, 0, sizeof(in_mem_A), &in_mem_A);
+        test_error(error, "clSetKernelArg failed");
+
+        error = clSetKernelArg(kernel, 1, sizeof(out_mem_A), &out_mem_A);
+        test_error(error, "clSetKernelArg failed");
+
+        return CL_SUCCESS;
+    }
+
+    cl_int Run() override
+    {
+        cl_int err = SetUp();
+        test_error(err, "SetUp failed");
+
+        const cl_int pattern_A = 42;
+        const cl_int pattern_B = 0xACDC;
+
+        {
+            clEventWrapper user_event = clCreateUserEvent(context, &err);
+            test_error(err, "clCreateUserEvent failed");
+
+            // enqueue producer operations
+            err = clEnqueueFillBuffer(
+                producer_queue, in_mem_A, &pattern_A, sizeof(cl_int), 0,
+                sizeof(cl_int) * num_elems, 1, &user_event, nullptr);
+            test_error(err, "clEnqueueReadBuffer failed");
+
+            err = clEnqueueFillBuffer(
+                producer_queue, in_mem_B, &pattern_B, sizeof(cl_int), 0,
+                sizeof(cl_int) * num_elems, 1, &user_event, nullptr);
+            test_error(err, "clEnqueueReadBuffer failed");
+
+            // launch producer operations simultaneously
+            err = clSetUserEventStatus(user_event, CL_COMPLETE);
+            test_error(err, "clSetUserEventStatus failed");
```

Comment:
This isn't a must-fix, but I don't think the user event is required here, since I don't think "real-world" code will hold back the producers like this.

Reply:
Corrected